---
title: "<img src='hex.png' width = 200><br>fluent genomic data analysis with R/Bioconductor"
author: "`r icon::fa_twitter()` _StuartLee <br> Monash University"
date: "2018-07-27"
output:
  xaringan::moon_reader:
    lib_dir: libs
    css: ["my.css", "libs/remark-css/default-fonts.css"]
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false

---

```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(comment = "")
library(plyranges)
```

class: center
# Bioconductor infrastructure is powerful...

<div style = "display:block; margin-left:auto; margin-right:auto">
  <img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/Bioconductor_1.0/Bioconductor.png", width = 200>
  <img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/GenomicRanges/GenomicRanges.png" width = "200">
  <img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/IRanges/IRanges.png" width = "200">
</div>

.large[
Bioconductor abstractions enable **developing** software for biology!
]

---

# ... however

```{r}
pkgs_to_get_started <- c("S4Vectors", "IRanges", "GenomicRanges")
pkg_classes <- function(.) methods::getClasses(asNamespace(.))

n_classes <- pkgs_to_get_started %>% 
  lapply(pkg_classes) %>% 
  lengths() %>% 
  sum()
n_classes

n_methods <- length(methods(class = "Ranges"))
n_methods
```

.large[A lot is expected of an R beginner to **analyse** data with Bioconductor]

---
class: inverse, center, middle

# How far can a user get with just a GRanges?

---

# GRanges are tidy!

```{r, fig.align='center', echo=FALSE}
knitr::include_graphics("GRanges.png")
```

---

# Design a grammar


```{r, echo = FALSE, fig.align = "center", out.width="1080"}

knitr::include_graphics("design.png")

```

---

# Include genomic semantics

.large[

- **arithmetic**: mutating/shifting/re-sizing/flanking
- **restriction**: filtering by metadata or ranges
- **aggregation**: summarizing over groups/overlaps/unions
- **merging**: combining ranges based on overlaps/nearest neighbors


_Biorxiv 327841_
]


---

## Grammar elements: arithimetic

.large[ The start, end, and width of a GRanges of are mutually dependent ]

```{r, eval = FALSE}
gr <- GRanges("chr1:10-20:+")

gr %>%
  mutate(width = width + 5L)
```

.large[How do we express changes to width, while fixing other coordinates?]

---

## Anchoring operator

An anchor decorates a GRanges and changes downstream behaviour 

```{r, eval = FALSE}
gr %>% 
  anchor_end() %>% 
  mutate(width = width + 5L)

gr %>% 
  anchor_5p() %>% 
  mutate(width = width + 5L)
```

---

## Grammar element: restrictions

.large[Similar to anchoring, grouping partitions a GRanges object and modifies any downstream behavior:]

```{r, eval = FALSE}
exons %>% 
  group_by(gene_id) %>% 
  filter(n() > 1)
```

.large[And include ability to restrict by overlaps:]
```{r, eval = FALSE}
exons %>%
  filter_by_non_overlaps(snps)
```

---
## Grammar elements: aggregation

.large[We want to summarise metadata over genomic coordinates or by groups (or both)]

```{r, eval = FALSE}
exons %>% 
  group_by(gene_id) %>% 
  summarise(score = sum(score))

exons %>% 
  group_by(gene_id) %>% 
  # merge overlapping regions in a gene
  reduce_ranges(score = sum(score))
```
---

## Grammar element: merging

.large[
- Reimagine overlap/nearest neighbour operations as table joins
- Always be explicit about strand
]

---

# Overlap Joins

```{r, echo = FALSE, fig.align='center'}
knitr::include_graphics("olap.png")
```

---

# Summary

.large[

- GRanges are tidy!!!

- A cognitive framework for playing with GRanges

- A stepping stone from tidyverse to Bioconductor

- Learn more at workshop today at 4pm in VC206
]

---

# Acknowledgements

- **Michael Lawrence**
- **Di Cook**
- Matt Ritchie
- Charity Law
- Shian Su
- Earo Wang

Development: https://github.com/sa-lee/plyranges

Contributions/feedback welcome from the Bioconductor community!
