---
title: "<img src='hex.png' width = 200><br>fluent genomic data analysis with R/Bioconductor"
author: "Stuart Lee<br>Monash University"
date: "2018-07-27"
output:
  xaringan::moon_reader:
    lib_dir: libs
    css: ["my.css", "libs/remark-css/default-fonts.css"]
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false

---

```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(comment = "")
library(plyranges)
```

class: center
# Bioconductor infrastructure is powerful...

<div style = "display:block; margin-left:auto; margin-right:auto">
  <img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/Bioconductor_1.0/Bioconductor.png", width = 200>
  <img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/GenomicRanges/GenomicRanges.png" width = "200">
  <img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/IRanges/IRanges.png" width = "200">
</div>

.large[
Bioconductor abstractions enable **developing** software for biology!
]

---

# ... however

```{r}
pkgs_to_get_started <- c("S4Vectors", "IRanges", "GenomicRanges")
pkg_classes <- function(.) methods::getClasses(asNamespace(.))

n_classes <- pkgs_to_get_started %>% 
  lapply(pkg_classes) %>% 
  lengths() %>% 
  sum()
n_classes

n_methods <- length(methods(class = "Ranges"))
n_methods
```

.large[A lot is expected of an R beginner to **analyse** data with Bioconductor]

---
class: inverse, center, middle

# How far can a user get with just a GRanges?

---

# GRanges are tidy!

```{r, fig.align='center', echo=FALSE}
knitr::include_graphics("GRanges.png")
```

---

# Design a grammar


```{r, echo = FALSE, fig.align = "center", out.width="1080"}

knitr::include_graphics("design.png")

```

---

# Include genomic semantics

.large[

- **arithmetic**: mutating/shifting/re-sizing/flanking
- **restriction**: filtering by metadata or ranges
- **aggregation**: summarizing over groups/overlaps/unions
- **merging**: combining ranges based on overlaps/nearest neighbors



_Biorxiv 327841_
]


---

# Finding introns in a non-model organism

.large[I have a GFF file from a non-model organism ðŸ˜‘.

How can I find introns?

What's our workflow?
]
 
---



## Reading in the GFF

```{r}
gff_rngs <- read_gff("my.gff")

gff_rngs %>% 
  select(ID, type) %>% 
  head(3)
```

---

## Retrieve exons and extract IDs

```{r}
all_exons <- gff_rngs %>% 
  filter(type == "exon") %>% 
  mutate(exon_rank = sub("(.*)-([^-]+$)", "\\2", ID),
         gene_id = gsub("exon_|-[^-]+$", "", ID),
         exon_id = paste0(gene_id, ".", exon_rank)) %>% 
  select(gene_id, exon_id, exon_rank) 

all_exons %>% 
  head(3)
```

---

## For each gene, merge overlapping exons

```{r}
sub_exons <- all_exons %>% 
  group_by(gene_id) %>%
  # this will merge overlapping exons for each group
  reduce_ranges_directed()

sub_exons %>% 
  head(3)
```
---

## Get the ranges for each gene from the exons

```{r}
genes <- all_exons %>% 
  group_by(seqnames, strand, gene_id) %>% 
  summarise(start = min(start), end = max(end)) %>% 
  as_granges()

# other, simpler way
# genes <- gff_rngs %>% filter(type == "gene")

genes %>% 
  head(3)
```

---


## Disjoin over each genes interval and filter

```{r}
introns <- genes %>% 
  bind_ranges(sub_exons) %>% 
  group_by(gene_id) %>% 
  disjoin_ranges_directed() %>% 
  filter_by_non_overlaps(sub_exons)

introns 
```

---

# Summary

.large[

- GRanges are tidy!!!

- A cognitive framework for playing with GRanges

- a stepping stone from tidyverse to Bioconductor

]

---

# Acknowledgements

- **Michael Lawrence**
- **Di Cook**
- Matt Ritchie
- Charity Law
- Shian Su
- Earo Wang

.large[
Development: https://github.com/sa-lee/plyranges

Contributions/feedback from the Bioconductor community!
]