<!DOCTYPE html>
<html>
  <head>
    <title> fluent genomic data analysis with R/Bioconductor</title>
    <meta charset="utf-8">
    <meta name="author" content="Stuart Lee Monash University" />
    <meta name="date" content="2018-07-27" />
    <link rel="stylesheet" href="my.css" type="text/css" />
    <link rel="stylesheet" href="libs/remark-css/default-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# <img src='hex.png' width = 200><br>fluent genomic data analysis with R/Bioconductor
### Stuart Lee<br>Monash University
### 2018-07-27

---




class: center
# Bioconductor infrastructure is powerful...

&lt;div style = "display:block; margin-left:auto; margin-right:auto"&gt;
  &lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/Bioconductor_1.0/Bioconductor.png", width = 200&gt;
  &lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/GenomicRanges/GenomicRanges.png" width = "200"&gt;
  &lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/IRanges/IRanges.png" width = "200"&gt;
&lt;/div&gt;
---

# ...but the learning curve is steep!


```r
pkgs_to_get_started &lt;- c("S4Vectors", "IRanges", "GenomicRanges")
pkg_classes &lt;- function(.) methods::getClasses(asNamespace(.))

n_classes &lt;- pkgs_to_get_started %&gt;% 
  lapply(pkg_classes) %&gt;% 
  lengths() %&gt;% 
  sum()
n_classes
```

```
[1] 143
```

```r
n_methods &lt;- length(methods(class = "Ranges"))
n_methods
```

```
[1] 210
```

---
class: inverse, center, middle

# How far can a user get with just a GRanges?


---

# GRanges are tidy!

&lt;img src="GRanges.png" style="display: block; margin: auto;" /&gt;

---

# Design a grammar


&lt;img src="design.png" width="1080" style="display: block; margin: auto;" /&gt;

---

# Include genomic semantics

.large[

- **arithmetic**: mutating/shifting/re-sizing/flanking
- **restriction**: filtering by metadata or ranges
- **aggregation**: summarizing over groups/overlaps/unions
- **merging**: combining ranges based on overlaps/nearest neighbors

]

See also: https://bit.ly/2v61UVs
---

# Finding introns in a non-model organism

.large[I have a GFF file from a non-model organism ðŸ˜‘.

How can I find introns?]
 
---

## Reading in the GFF


```r
gff_rngs &lt;- read_gff("my.gff")

gff_rngs %&gt;% 
  select(ID, type) %&gt;% 
  head(3)
```

```
GRanges object with 3 ranges and 2 metadata columns:
         seqnames        ranges strand |                    ID     type
            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;character&gt; &lt;factor&gt;
  [1] PvP01_11_v1 634386-635717      + |         PVP01_1114900     gene
  [2] PvP01_11_v1 634386-635717      + |       PVP01_1114900.1     mRNA
  [3] PvP01_11_v1 634386-635717      + | exon_PVP01_1114900-E1     exon
  -------
  seqinfo: 178 sequences from an unspecified genome; no seqlengths
```

---

## Retrieve exons and extract IDs


```r
all_exons &lt;- gff_rngs %&gt;% 
  filter(type == "exon") %&gt;% 
  mutate(exon_rank = sub("(.*)-([^-]+$)", "\\2", ID),
         gene_id = gsub("exon_|-[^-]+$", "", ID),
         exon_id = paste0(gene_id, ".", exon_rank)) %&gt;% 
  select(gene_id, exon_id, exon_rank) 

all_exons %&gt;% 
  head(3)
```

```
GRanges object with 3 ranges and 3 metadata columns:
         seqnames        ranges strand |       gene_id          exon_id
            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;character&gt;      &lt;character&gt;
  [1] PvP01_11_v1 634386-635717      + | PVP01_1114900 PVP01_1114900.E1
  [2] PvP01_07_v1 348842-348940      - | PVP01_0706500 PVP01_0706500.E3
  [3] PvP01_07_v1 349108-349199      - | PVP01_0706500 PVP01_0706500.E2
        exon_rank
      &lt;character&gt;
  [1]          E1
  [2]          E3
  [3]          E2
  -------
  seqinfo: 178 sequences from an unspecified genome; no seqlengths
```

---

## For each gene, merge overlapping exons


```r
sub_exons &lt;- all_exons %&gt;% 
  group_by(gene_id) %&gt;%
  # this will merge overlapping exons for each group
  reduce_ranges_directed()

sub_exons %&gt;% 
  head(3)
```

```
GRanges object with 3 ranges and 1 metadata column:
                       seqnames    ranges strand |       gene_id
                          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;character&gt;
  [1] Transfer.PvP01_00_1.final    53-135      + | PVP01_0000010
  [2] Transfer.PvP01_00_1.final   199-415      + | PVP01_0000010
  [3] Transfer.PvP01_00_1.final   502-669      + | PVP01_0000010
  -------
  seqinfo: 178 sequences from an unspecified genome; no seqlengths
```
---

## Get the ranges for each gene from the exons


```r
genes &lt;- all_exons %&gt;% 
  group_by(seqnames, strand, gene_id) %&gt;% 
  summarise(start = min(start), end = max(end)) %&gt;% 
  as_granges()

# other, simpler way
# genes &lt;- gff_rngs %&gt;% filter(type == "gene")

genes %&gt;% 
  head(3)
```

```
GRanges object with 3 ranges and 1 metadata column:
                       seqnames     ranges strand |       gene_id
                          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;character&gt;
  [1] Transfer.PvP01_00_1.final     53-669      + | PVP01_0000010
  [2] Transfer.PvP01_00_1.final  2797-3513      + | PVP01_0000020
  [3] Transfer.PvP01_00_1.final 7307-10065      + | PVP01_0000030
  -------
  seqinfo: 178 sequences from an unspecified genome; no seqlengths
```

---

## Combine the genes and exon ranges, and group by gene


```r
introns &lt;- genes %&gt;% 
  bind_ranges(sub_exons) %&gt;% 
  group_by(gene_id)

introns
```

```
GRanges object with 24584 ranges and 1 metadata column:
Groups: gene_id [6830]
                           seqnames      ranges strand |        gene_id
                              &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt;
      [1] Transfer.PvP01_00_1.final      53-669      + |  PVP01_0000010
      [2] Transfer.PvP01_00_1.final   2797-3513      + |  PVP01_0000020
      [3] Transfer.PvP01_00_1.final  7307-10065      + |  PVP01_0000030
      [4] Transfer.PvP01_00_1.final 10224-12567      + |  PVP01_0000040
      [5] Transfer.PvP01_00_1.final 14215-16428      - |  PVP01_0000050
      ...                       ...         ...    ... .            ...
  [24580]              PvP01_MIT_v1   3066-3101      + | PVP01_MIT03100
  [24581]              PvP01_MIT_v1   3102-3182      + | PVP01_MIT03200
  [24582]              PvP01_MIT_v1   3267-4760      + | PVP01_MIT03300
  [24583]              PvP01_MIT_v1   4782-5912      + | PVP01_MIT03400
  [24584]              PvP01_MIT_v1   5914-5985      + | PVP01_MIT03500
  -------
  seqinfo: 178 sequences from an unspecified genome; no seqlengths
```

---

## Disjoin over each genes interval and filter


```r
introns &lt;- introns %&gt;% 
  disjoin_ranges_directed() %&gt;% 
  filter_by_non_overlaps(sub_exons)

introns 
```

```
GRanges object with 10913 ranges and 1 metadata column:
                           seqnames          ranges strand |
                              &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |
      [1] Transfer.PvP01_00_1.final         136-198      + |
      [2] Transfer.PvP01_00_1.final         416-501      + |
      [3] Transfer.PvP01_00_1.final       7993-8048      + |
      [4] Transfer.PvP01_00_1.final       8714-8845      + |
      [5] Transfer.PvP01_00_1.final     11227-11285      + |
      ...                       ...             ...    ... .
  [10909]               PvP01_14_v1 3143386-3143490      - |
  [10910]               PvP01_14_v1 3144211-3144350      - |
  [10911]               PvP01_14_v1 3146171-3146278      - |
  [10912]               PvP01_14_v1 3147482-3147623      - |
  [10913]              PvP01_API_v1     15883-15884      - |
                 gene_id
             &lt;character&gt;
      [1]  PVP01_0000010
      [2]  PVP01_0000010
      [3]  PVP01_0000030
      [4]  PVP01_0000030
      [5]  PVP01_0000040
      ...            ...
  [10909]  PVP01_1472900
  [10910]  PVP01_1472900
  [10911]  PVP01_1473000
  [10912]  PVP01_1473000
  [10913] PVP01_API04100
  -------
  seqinfo: 178 sequences from an unspecified genome; no seqlengths
```

---

# Summary

.large[

- GRanges are tidy!!!

- A cognitive framework for playing with GRanges

- a stepping stone from tidyverse to biocverse 

]

---

# Acknowledgements

- **Michael Lawrence**
- **Di Cook**
- Matt Ritchie
- Charity Law
- Shian Su
- Earo Wang

.large[
Development: https://github.com/sa-lee/plyranges

Contributions/feedback from the Bioconductor community!
]
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script>
(function() {
  var i, text, code, codes = document.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
})();
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
