<!DOCTYPE html>
<html>
  <head>
    <title> fluent genomic data analysis with R/Bioconductor</title>
    <meta charset="utf-8">
    <meta name="author" content="<i class="fab  fa-twitter "></i> _StuartLee   Monash University" />
    <meta name="date" content="2018-07-27" />
    <link href="libs/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="my.css" type="text/css" />
    <link rel="stylesheet" href="libs/remark-css/default-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# <img src='hex.png' width = 200><br>fluent genomic data analysis with R/Bioconductor
### <i class="fab  fa-twitter "></i> _StuartLee <br> Monash University
### 2018-07-27

---




class: center
# Bioconductor infrastructure is powerful...

&lt;div style = "display:block; margin-left:auto; margin-right:auto"&gt;
  &lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/Bioconductor_1.0/Bioconductor.png", width = 200&gt;
  &lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/GenomicRanges/GenomicRanges.png" width = "200"&gt;
  &lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/master/IRanges/IRanges.png" width = "200"&gt;
&lt;/div&gt;

.large[
Bioconductor abstractions enable **developing** software for biology!
]

---

# ... however


```r
pkgs_to_get_started &lt;- c("S4Vectors", "IRanges", "GenomicRanges")
pkg_classes &lt;- function(.) methods::getClasses(asNamespace(.))

n_classes &lt;- pkgs_to_get_started %&gt;% 
  lapply(pkg_classes) %&gt;% 
  lengths() %&gt;% 
  sum()
n_classes
```

```
[1] 143
```

```r
n_methods &lt;- length(methods(class = "Ranges"))
n_methods
```

```
[1] 210
```

.large[A lot is expected of an R beginner to **analyse** data with Bioconductor]

---
class: inverse, center, middle

# How far can a user get with just a GRanges?

---

# GRanges are tidy!

&lt;img src="GRanges.png" style="display: block; margin: auto;" /&gt;

---

# Design a grammar


&lt;img src="design.png" width="1080" style="display: block; margin: auto;" /&gt;

---

# Include genomic semantics

.large[

- **arithmetic**: mutating/shifting/re-sizing/flanking
- **restriction**: filtering by metadata or ranges
- **aggregation**: summarizing over groups/overlaps/unions
- **merging**: combining ranges based on overlaps/nearest neighbors


_Biorxiv 327841_
]


---

## Grammar elements: arithimetic

.large[ The start, end, and width of a GRanges of are mutually dependent ]


```r
gr &lt;- GRanges("chr1:10-20:+")

gr %&gt;%
  mutate(width = width + 5L)
```

.large[How do we express changes to width, while fixing other coordinates?]

---

## Anchoring operator

An anchor decorates a GRanges and changes downstream behaviour 


```r
gr %&gt;% 
  anchor_end() %&gt;% 
  mutate(width = width + 5L)

gr %&gt;% 
  anchor_5p() %&gt;% 
  mutate(width = width + 5L)
```

---

## Grammar element: restrictions

.large[Similar to anchoring, grouping partitions a GRanges object and modifies any downstream behavior:]


```r
exons %&gt;% 
  group_by(gene_id) %&gt;% 
  filter(n() &gt; 1)
```

.large[And include ability to restrict by overlaps:]

```r
exons %&gt;%
  filter_by_non_overlaps(snps)
```

---
## Grammar elements: aggregation

.large[We want to summarise metadata over genomic coordinates or by groups (or both)]


```r
exons %&gt;% 
  group_by(gene_id) %&gt;% 
  summarise(score = sum(score))

exons %&gt;% 
  group_by(gene_id) %&gt;% 
  # merge overlapping regions in a gene
  reduce_ranges(score = sum(score))
```
---

## Grammar element: merging

.large[
- Reimagine overlap/nearest neighbour operations as table joins
- Always be explicit about strand
]

---

# Overlap Joins

&lt;img src="olap.png" style="display: block; margin: auto;" /&gt;

---

# Summary

.large[

- GRanges are tidy!!!

- A cognitive framework for playing with GRanges

- A stepping stone from tidyverse to Bioconductor

- Learn more at workshop today at 4pm in VC206
]

---

# Acknowledgements

- **Michael Lawrence**
- **Di Cook**
- Matt Ritchie
- Charity Law
- Shian Su
- Earo Wang

Development: https://github.com/sa-lee/plyranges

Contributions/feedback welcome from the Bioconductor community!
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script>
(function() {
  var i, text, code, codes = document.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
})();
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
