---
title: "plyranges"
subtitle: "fluent genomic data analysis with R/Bioconductor"
author: "Stuart Lee"
date: "2018-07-27"
output:
  xaringan::moon_reader:
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---

```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(comment = "#>")
library(magrittr)
library(plyranges)
```

## Bioconductor infrastructure is powerful...

<!-- include package hex stickers -->

---

## but the learning curve is steep

```{r}
pkgs_to_get_started <- c("S4Vectors", "IRanges", "GenomicRanges")
pkg_classes <- function(.) methods::getClasses(asNamespace(.))

n_classes <- pkgs_to_get_started %>% 
  lapply(pkg_classes) %>% 
  lengths() %>% 
  sum()
n_classes

n_methods <- showMethods(classes = "Ranges",  printTo = FALSE) %>% 
  grep("^Function", .) %>% 
  length()
n_methods
```

---
class: inverse, center, middle

# How far can we get with just a GRanges?


---

# GRanges are tidy...

```{r, fig.align='center', echo=FALSE}
knitr::include_graphics("GRanges.png")
```


---

# ... so we can design a grammar

<!-- maybe a figure would be better here -->


- functions are verbs
- GRanges in, GRanges out
- leverage existing Bioconductor infrastructure 
- enable a user to develop expressive workflows

---

# Finding introns in a non-model organism

I have a GFF file from a non-model organism ðŸ˜‘.

How can I find introns by (single) genes?

---

## Reading in the GFF

```{r}

gff_rngs <- read_gff("PlasmoDB-38_PvivaxP01.gff")
head(gff_rngs)
```

A mess!
---

## Retrieve exons and extract IDs

```{r}
all_exons <- gff_rngs %>% 
  filter(type == "exon") %>% 
  mutate(exon_rank = sub("(.*)-([^-]+$)", "\\2", ID),
         gene_id = gsub("exon_|-[^-]+$", "", ID),
         exon_id = paste0(gene_id, ".", exon_rank)) %>% 
  select(gene_id, exon_id, exon_rank) 
head(all_exons)
```

---

## Merge overlapping exons within a gene

```{r}
sub_exons <- all_exons %>% 
  group_by(gene_id) %>% 
  reduce_ranges_directed() %>% 
  group_by(gene_id) %>% 
  filter(n() > 1) %>% 
  ungroup() 

head(sub_exons)
```

---

## Subtract exonic regions from genes

```{r}
genes <- gff_rngs %>% 
  filter(type == "gene") %>% 
  mutate(gene_id = ID) %>% 
  select(gene_id)

introns <- genes %>% 
  bind_ranges(sub_exons) %>% 
  group_by(gene_id) %>% 
  disjoin_ranges_directed() %>% 
  filter_by_non_overlaps(all_exons)
head(introns)
```


---

## Summary

```{r, echo = FALSE, fig.align='center', out.width='600'}
knitr::include_graphics("hex.png")
```

---

# Acknowledgements

- **Michael Lawrence**
- **Di Cook**
- Matt Ritchie
- Charity Law
- Shian Su
- Earo Wang
